File: src-tauri/src/backend/svg_row.rs
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;
use serde::{Serialize, Serializer};

#[derive(Clone)]
pub enum SVGRowPropertyAttrs {
    SomeString(String),
    SomeInt(isize),
}

impl Serialize for SVGRowPropertyAttrs {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> where S: Serializer {
        match &self {
            SVGRowPropertyAttrs::SomeString(st) => st.serialize(serializer),
            SVGRowPropertyAttrs::SomeInt(i) => i.serialize(serializer),
        }
    }
}

#[derive(Clone)]
pub enum SVGRowProperty {
    SomeInt(isize),
    SomeString(String),
    SomeHashMap(HashMap<String, SVGRowPropertyAttrs>),
}

impl Serialize for SVGRowProperty {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> where S: Serializer {
        match &self {
            SVGRowProperty::SomeInt(i) => i.serialize(serializer),
            SVGRowProperty::SomeString(st) => st.serialize(serializer),
            SVGRowProperty::SomeHashMap(hm) => hm.serialize(serializer),
        }
    }
}

#[derive(Clone)]
pub enum DrawProperty {
    SomeHashMap(HashMap<String, SVGRowProperty>),
    SomeVector(Vec<HashMap<String, SVGRowProperty>>),
    SomeVectorVector(Vec<Vec<HashMap<String, SVGRowProperty>>>),
}

impl Serialize for DrawProperty {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> where S: Serializer {
        match &self {
            DrawProperty::SomeHashMap(hm) => hm.serialize(serializer),
            DrawProperty::SomeVector(v) => v.serialize(serializer),
            DrawProperty::SomeVectorVector(v) => v.serialize(serializer),
        }
    }
}

const Y_OFFSET: isize = 20;
const X_OFFSET: isize = 20;  // If changing, be sure to update on front-end too
const X_SPACING: isize = 20;  // If changing, be sure to update on front-end too
const Y_SPACING: isize = 30;
const TEXT_Y_ALIGNMENT: isize = 6;
const CIRCLE_RADIUS: isize = 10;
const RECT_Y_OFFSET: isize = -12;
const RECT_HEIGHT: isize = 24;

#[derive(Clone)]
pub struct SVGRow {
    sha: String,
    summary: String,
    branches_and_tags: Vec<(String, String)>,
    parent_oids: Vec<String>,
    child_oids: Vec<String>,
    x: isize,
    y: isize,
}

impl SVGRow {
    pub fn new(sha: String, summary: String, branches_and_tags: Vec<(String, String)>, parent_oids: Vec<String>, child_oids: Vec<String>, x: isize, y: isize) -> Self {
        Self {
            sha,
            summary,
            branches_and_tags,
            parent_oids,
            child_oids,
            x,
            y,
        }
    }

    pub fn get_parent_or_child_svg_row_values(&self, all_svg_rows: &HashMap<String, Rc<RefCell<SVGRow>>>, sha_type: String) -> Result<Vec<Rc<RefCell<SVGRow>>>, Box<dyn std::error::Error>> {
        let mut svg_row_values: Vec<Rc<RefCell<SVGRow>>> = vec![];
        let shas;
        if sha_type == "parents" {
            shas = &self.parent_oids;
        } else if sha_type == "children" {
            shas = &self.child_oids;
        } else {
            return Err("Please use parents or children for sha_type.".into());
        }
        for sha in shas {
            match all_svg_rows.get(&*sha) {
                Some(svg_row_rc) => {
                    svg_row_values.push(svg_row_rc.clone());
                },
                // If a parent or child is not present, ignore it. It may be outside the revwalk range.
                None => (),
            };
        }
        Ok(svg_row_values)
    }

    fn get_color_string(x: isize) -> String {
        let color_num = x % 4;
        if color_num == 0 {
            String::from("#00CC19")
        } else if color_num == 1 {
            String::from("#0198A6")
        } else if color_num == 2 {
            String::from("#FF7800")
        } else {
            String::from("#FF0D00")
        }
    }

    pub fn get_draw_properties(&mut self, main_table: &mut HashMap<isize, HashMap<isize, bool>>, parent_svg_rows: Vec<Rc<RefCell<SVGRow>>>, child_svg_rows: Vec<Rc<RefCell<SVGRow>>>) -> Vec<DrawProperty> {
        // Set the current node position as occupied (or find a position that's unoccupied and occupy it).
        match main_table.get_mut(&self.y) {
            Some(hm) => {
                match hm.get(&self.x) {
                    Some(is_occupied) => {
                        if *is_occupied == true {
                            let mut found_empty = false;
                            while !found_empty {
                                self.x += 1;
                                if !hm.contains_key(&self.x) {
                                    found_empty = true;
                                    hm.insert(self.x, true);
                                }
                            }
                        }
                    },
                    None => {
                        hm.insert(self.x, true);
                    },
                };
            },
            None => {
                let mut temp_hm: HashMap<isize, bool> = HashMap::new();
                temp_hm.insert(self.x, true);
                main_table.insert(self.y, temp_hm);
            },
        };

        // Set the space of the line from the current node to its parents as occupied.
        for parent_svg_row in parent_svg_rows {
            for i in (self.y + 1)..parent_svg_row.borrow().y {
                match main_table.get_mut(&i) {
                    Some(hm) => {
                        if !hm.contains_key(&self.x) {
                            hm.insert(self.x, true);
                        }
                    },
                    None => {
                        let mut temp_hm: HashMap<isize, bool> = HashMap::new();
                        temp_hm.insert(self.x, true);
                        main_table.insert(i, temp_hm);
                    },
                };
            }
        }

        let mut draw_properties: Vec<DrawProperty> = vec![];

        let pixel_x = self.x * X_SPACING + X_OFFSET;
        let pixel_y = self.y * Y_SPACING + Y_OFFSET;
        let color = SVGRow::get_color_string(self.x);
        let mut child_lines: Vec<HashMap<String, SVGRowProperty>> = vec![];
        // Draw the lines from the current node to its children.
        for child_svg_row in child_svg_rows {
            let child_svg_row_b = child_svg_row.borrow();
            let child_pixel_x = child_svg_row_b.x * X_SPACING + X_OFFSET;
            let child_pixel_y = child_svg_row_b.y * Y_SPACING + Y_OFFSET;
            let before_y = self.y - 1;
            let before_pixel_y = before_y * Y_SPACING + Y_OFFSET;
            if before_pixel_y != child_pixel_y {
                for i in child_svg_row_b.y..before_y {
                    let top_pixel_y = i * Y_SPACING + Y_OFFSET;
                    let bottom_pixel_y = (i + 1) * Y_SPACING + Y_OFFSET;

                    let mut style_str = String::from("stroke:");
                    style_str.push_str(&*SVGRow::get_color_string(child_svg_row_b.x));
                    style_str.push_str(";stroke-width:4");
                    let line_attrs: HashMap<String, SVGRowPropertyAttrs> = HashMap::from([
                        (String::from("x1"), SVGRowPropertyAttrs::SomeInt(child_pixel_x)),
                        (String::from("y1"), SVGRowPropertyAttrs::SomeInt(top_pixel_y)),
                        (String::from("x2"), SVGRowPropertyAttrs::SomeInt(child_pixel_x)),
                        (String::from("y2"), SVGRowPropertyAttrs::SomeInt(bottom_pixel_y)),
                        (String::from("style"), SVGRowPropertyAttrs::SomeString(style_str)),
                    ]);
                    child_lines.push(HashMap::from([
                        (String::from("tag"), SVGRowProperty::SomeString(String::from("line"))),
                        (String::from("attrs"), SVGRowProperty::SomeHashMap(line_attrs)),
                        (String::from("row-y"), SVGRowProperty::SomeInt(i + 1)),
                    ]));
                }
            }
            let mut style_str = String::from("stroke:");
            let row_y;
            if child_svg_row_b.x >= self.x {
                // Sets the color for "branching" lines and straight lines
                style_str.push_str(&*SVGRow::get_color_string(child_svg_row_b.x));
                row_y = self.y;
            } else {
                // Sets the color for "merging" lines
                style_str.push_str(&*SVGRow::get_color_string(self.x));
                row_y = before_y;
            }
            style_str.push_str(";fill:transparent;stroke-width:4");
            if child_pixel_x == pixel_x {
                let line_attrs: HashMap<String, SVGRowPropertyAttrs> = HashMap::from([
                    (String::from("x1"), SVGRowPropertyAttrs::SomeInt(child_pixel_x)),
                    (String::from("y1"), SVGRowPropertyAttrs::SomeInt(before_pixel_y)),
                    (String::from("x2"), SVGRowPropertyAttrs::SomeInt(pixel_x)),
                    (String::from("y2"), SVGRowPropertyAttrs::SomeInt(pixel_y)),
                    (String::from("style"), SVGRowPropertyAttrs::SomeString(style_str)),
                ]);
                child_lines.push(HashMap::from([
                    (String::from("tag"), SVGRowProperty::SomeString(String::from("line"))),
                    (String::from("attrs"), SVGRowProperty::SomeHashMap(line_attrs)),
                    (String::from("row-y"), SVGRowProperty::SomeInt(row_y)),
                ]));
            } else {
                let mut d_str = format!("M {child_pixel_x} {before_pixel_y} C ");
                if child_pixel_x < pixel_x {
                    let start_control_point_x = child_pixel_x + X_SPACING * 3 / 4;
                    let end_control_point_y = pixel_y - Y_SPACING * 3 / 4;
                    d_str.push_str(&*format!("{start_control_point_x} {before_pixel_y}, {pixel_x} {end_control_point_y}, "));
                } else {
                    let start_control_point_y = before_pixel_y + Y_SPACING * 3 / 4;
                    let end_control_point_x = pixel_x + X_SPACING * 3 / 4;
                    d_str.push_str(&*format!("{child_pixel_x} {start_control_point_y}, {end_control_point_x} {pixel_y}, "));
                }
                d_str.push_str(&*format!("{pixel_x} {pixel_y}"));
                let path_attrs: HashMap<String, SVGRowPropertyAttrs> = HashMap::from([
                    (String::from("d"), SVGRowPropertyAttrs::SomeString(d_str)),
                    (String::from("style"), SVGRowPropertyAttrs::SomeString(style_str)),
                ]);
                child_lines.push(HashMap::from([
                    (String::from("tag"), SVGRowProperty::SomeString(String::from("path"))),
                    (String::from("attrs"), SVGRowProperty::SomeHashMap(path_attrs)),
                    (String::from("row-y"), SVGRowProperty::SomeInt(row_y)),
                ]));
            }
        }
        draw_properties.push(DrawProperty::SomeVector(child_lines));

        // Now get the circle
        let circle_attrs: HashMap<String, SVGRowPropertyAttrs> = HashMap::from([
            (String::from("cx"), SVGRowPropertyAttrs::SomeInt(pixel_x)),
            (String::from("cy"), SVGRowPropertyAttrs::SomeInt(pixel_y)),
            (String::from("r"), SVGRowPropertyAttrs::SomeInt(CIRCLE_RADIUS)),
            (String::from("stroke"), SVGRowPropertyAttrs::SomeString(color.clone())),
            (String::from("stroke-width"), SVGRowPropertyAttrs::SomeInt(1)),
            (String::from("fill"), SVGRowPropertyAttrs::SomeString(color.clone())),
        ]);
        draw_properties.push(DrawProperty::SomeHashMap(HashMap::from([
            (String::from("tag"), SVGRowProperty::SomeString(String::from("circle"))),
            (String::from("attrs"), SVGRowProperty::SomeHashMap(circle_attrs)),
        ])));

        // Get the branch text
        let empty_hm = HashMap::new();
        let largest_occupied_x = main_table.get(&self.y).unwrap_or(&empty_hm).keys().max().unwrap_or(&0);
        let mut branch_and_tags: Vec<Vec<HashMap<String, SVGRowProperty>>> = vec![];
        for (branch_name, branch_type) in self.branches_and_tags.clone().into_iter() {
            let mut branch_and_tag_properties: Vec<HashMap<String, SVGRowProperty>> = vec![];
            let text_attrs: HashMap<String, SVGRowPropertyAttrs> = HashMap::from([
                (String::from("x"), SVGRowPropertyAttrs::SomeInt(0)),
                (String::from("y"), SVGRowPropertyAttrs::SomeInt(pixel_y + TEXT_Y_ALIGNMENT)),
                (String::from("fill"), SVGRowPropertyAttrs::SomeString(String::from("white"))),
            ]);
            branch_and_tag_properties.push(HashMap::from([
                (String::from("tag"), SVGRowProperty::SomeString(String::from("text"))),
                (String::from("attrs"), SVGRowProperty::SomeHashMap(text_attrs)),
                (String::from("textContent"), SVGRowProperty::SomeString(branch_name.clone())),
                (String::from("largestXValue"), SVGRowProperty::SomeInt(*largest_occupied_x)),
            ]));

            let mut branch_rect_color = "yellow";
            if branch_type == "local" {
                branch_rect_color = "red";
            } else if branch_type == "remote" {
                branch_rect_color = "green";
            } else if branch_type == "tag" {
                branch_rect_color = "grey";
            }

            let mut style_str = String::from("fill:");
            style_str.push_str(branch_rect_color);
            style_str.push_str(";fill-opacity:0.5;");
            let rect_attrs: HashMap<String, SVGRowPropertyAttrs> = HashMap::from([
                (String::from("x"), SVGRowPropertyAttrs::SomeInt(0)),
                (String::from("y"), SVGRowPropertyAttrs::SomeInt(pixel_y + RECT_Y_OFFSET)),
                (String::from("rx"), SVGRowPropertyAttrs::SomeInt(10)),
                (String::from("ry"), SVGRowPropertyAttrs::SomeInt(10)),
                (String::from("width"), SVGRowPropertyAttrs::SomeInt(0)),
                (String::from("height"), SVGRowPropertyAttrs::SomeInt(RECT_HEIGHT)),
                (String::from("style"), SVGRowPropertyAttrs::SomeString(style_str)),
            ]);
            branch_and_tag_properties.push(HashMap::from([
                (String::from("tag"), SVGRowProperty::SomeString(String::from("rect"))),
                (String::from("attrs"), SVGRowProperty::SomeHashMap(rect_attrs)),
            ]));
            branch_and_tags.push(branch_and_tag_properties);
        }
        draw_properties.push(DrawProperty::SomeVectorVector(branch_and_tags));

        // Get summary text
        let text_attrs: HashMap<String, SVGRowPropertyAttrs> = HashMap::from([
            (String::from("x"), SVGRowPropertyAttrs::SomeInt(0)),
            (String::from("y"), SVGRowPropertyAttrs::SomeInt(pixel_y + TEXT_Y_ALIGNMENT)),
            (String::from("fill"), SVGRowPropertyAttrs::SomeString(String::from("white"))),
        ]);
        draw_properties.push(DrawProperty::SomeHashMap(HashMap::from([
            (String::from("tag"), SVGRowProperty::SomeString(String::from("text"))),
            (String::from("attrs"), SVGRowProperty::SomeHashMap(text_attrs)),
            (String::from("textContent"), SVGRowProperty::SomeString(self.summary.clone())),
            (String::from("largestXValue"), SVGRowProperty::SomeInt(*largest_occupied_x)),
        ])));

        // Get background rectangle
        let rect_attrs: HashMap<String, SVGRowPropertyAttrs> = HashMap::from([
            (String::from("class"), SVGRowPropertyAttrs::SomeString(String::from("backgroundRect"))),
            (String::from("x"), SVGRowPropertyAttrs::SomeInt(pixel_x)),
            (String::from("y"), SVGRowPropertyAttrs::SomeInt(pixel_y + RECT_Y_OFFSET)),
            (String::from("width"), SVGRowPropertyAttrs::SomeInt(0)),
            (String::from("height"), SVGRowPropertyAttrs::SomeInt(RECT_HEIGHT)),
            (String::from("style"), SVGRowPropertyAttrs::SomeString(String::from("fill:white;fill-opacity:0.1;"))),
        ]);
        draw_properties.push(DrawProperty::SomeHashMap(HashMap::from([
            (String::from("tag"), SVGRowProperty::SomeString(String::from("rect"))),
            (String::from("attrs"), SVGRowProperty::SomeHashMap(rect_attrs)),
        ])));

        draw_properties
    }
}

File: ui/svg_row.js
/**
 * A row in the svg table.
 */
import {writeText} from '@tauri-apps/api/clipboard';

export class SVGRow {
    X_OFFSET = 20;
    Y_OFFSET = 20;
    X_SPACING = 20;
    Y_SPACING = 30;
    TEXT_Y_ALIGNMENT = 6;
    CIRCLE_RADIUS = 10;
    RECT_Y_OFFSET = -12;
    RECT_HEIGHT = 24;
    BRANCH_TEXT_SPACING = 5;

    /**
     * Construct the svg row
     * @param {string} sha
     * @param {string} summary
     * @param {Array<string>} branchesAndTags
     * @param {Array<string>} parentShas
     * @param {Array<string>} childrenShas
     * @param {int} x
     * @param {int} y
     */
    constructor(sha, summary, branchesAndTags, parentShas, childrenShas, x, y) {
        this.sha = sha;
        this.summary = summary;
        this.branchesAndTags = branchesAndTags;
        this.parentShas = parentShas;
        this.childrenShas = childrenShas;
        this.x = x;
        this.y = y;
        this.width = 0;
    }

    /**
     * Gets the SVGRow parents
     * @param {Array<SVGRow>} array
     * @return {Array<SVGRow>}
     */
    getParentSVGRows(array) {
        const self = this;
        if (self.parentShas.length === 0) {
            return [];
        }
        const parentSVGRows = [];
        for (let i = 0; i < self.parentShas.length; i++) {
            for (let j = 0; j < array.length; j++) {
                if (self.parentShas[i] === array[j].sha) {
                    parentSVGRows.push(array[j]);
                    break;
                }
            }
        }
        return parentSVGRows;
    }

    /**
     * Gets the SVGRow children
     * @param {Array<SVGRow>} array
     * @return {Array<SVGRow>}
     */
    getChildSVGRows(array) {
        const self = this;
        if (self.childrenShas.length === 0) {
            return [];
        }
        const childSVGRows = [];
        for (let i = 0; i < self.childrenShas.length; i++) {
            for (let j = 0; j < array.length; j++) {
                if (self.childrenShas[i] === array[j].sha) {
                    childSVGRows.push(array[j]);
                    break;
                }
            }
        }
        return childSVGRows;
    }

    /**
     * Draw each of the components of the svg row.
     * @param {jQuery} $commitTableSVG
     * @param {Array<SVGRow>} parentSVGRows
     * @param {Array<SVGRow>} childSVGRows
     * @param {Object<number, Object<number, boolean>>} mainTable
     */
    draw($commitTableSVG, parentSVGRows, childSVGRows, mainTable) {
        const self = this;

        // Set the current node position as occupied (or find a position that's unoccupied and occupy it).
        if (!(self.y in mainTable)) {
            mainTable[self.y] = {};
            mainTable[self.y][self.x] = true;
        } else if (!(self.x in mainTable[self.y])) {
            mainTable[self.y][self.x] = true;
        } else if (mainTable[self.y][self.x] === true) {
            let foundEmpty = false;
            while (!foundEmpty) {
                self.x++;
                if (!(self.x in mainTable[self.y])) {
                    foundEmpty = true;
                    mainTable[self.y][self.x] = true;
                }
            }
        }

        // Set the space of the line from the current node to its parents as occupied.
        const pixelX = self.x * self.X_SPACING + self.X_OFFSET;
        const pixelY = self.y * self.Y_SPACING + self.Y_OFFSET;
        const color = self.getColor(self.x);
        if (parentSVGRows.length > 0) {
            for (let i = 0; i < parentSVGRows.length; i++) {
                // ParentSVGRows are lower on the graph (with a higher y value).
                for (let j = self.y + 1; j < parentSVGRows[i].y; j++) {
                    if (!(j in mainTable)) {
                        mainTable[j] = {};
                        mainTable[j][self.x] = true;
                    } else if (!(self.x in mainTable[j])) {
                        mainTable[j][self.x] = true;
                    }
                }
            }
        }

        // Draw the lines from the current node to its children.
        if (childSVGRows.length > 0) {
            for (let i = 0; i < childSVGRows.length; i++) {
                const childPixelX = childSVGRows[i].x * self.X_SPACING + self.X_OFFSET;
                const childPixelY = childSVGRows[i].y * self.Y_SPACING + self.Y_OFFSET;
                const beforePixelY = (self.y - 1) * self.Y_SPACING + self.Y_OFFSET;
                const svgLine = self.makeSVG('line', {x1: childPixelX, y1: childPixelY, x2: childPixelX, y2: beforePixelY, style: 'stroke:' + self.getColor(childSVGRows[i].x) + ';stroke-width:4'});
                const angledSVGLine = self.makeSVG('line', {x1: childPixelX, y1: beforePixelY, x2: pixelX, y2: pixelY, style: 'stroke:' + self.getColor(childSVGRows[i].x) + ';stroke-width:4'});
                $commitTableSVG.append(svgLine);
                $commitTableSVG.append(angledSVGLine);
            }
        }

        // Now draw the node.
        const svgCircle = self.makeSVG('circle', {'cx': pixelX, 'cy': pixelY, 'r': self.CIRCLE_RADIUS, 'stroke': color, 'stroke-width': 1, 'fill': color});
        $commitTableSVG.append(svgCircle);

        // Draw the branch text.
        const occupiedRowNums = Object.keys(mainTable[self.y]);
        let largestXValue = 0;
        for (let i = 0; i < occupiedRowNums.length; i++) {
            largestXValue = Math.max(largestXValue, Number(occupiedRowNums[i]));
        }
        let currentX = (largestXValue + 1) * self.X_SPACING + self.X_OFFSET;
        const contextFunction = self.getContextFunction();
        for (let i = 0; i < self.branchesAndTags.length; i++) {
            const svgTextElem = self.makeSVG('text', {x: currentX, y: pixelY + self.TEXT_Y_ALIGNMENT, fill: 'white'});
            svgTextElem.textContent = self.branchesAndTags[i]['branch_name'];
            svgTextElem.oncontextmenu = contextFunction;
            const branchRectId = 'branch_rect_' + i + '_' + self.sha;
            let branchRectColor = 'yellow';
            if (self.branchesAndTags[i]['branch_type'] === 'local') {
                branchRectColor = 'red';
            } else if (self.branchesAndTags[i]['branch_type'] === 'remote') {
                branchRectColor = 'green';
            } else if (self.branchesAndTags[i]['branch_type'] === 'tag') {
                branchRectColor = 'grey';
            }
            const svgRectElem = self.makeSVG('rect', {id: branchRectId, x: currentX - 5, y: pixelY + self.RECT_Y_OFFSET, rx: 10, ry: 10, width: 0, height: self.RECT_HEIGHT, style: 'fill:' + branchRectColor + ';fill-opacity:0.5;'});
            $commitTableSVG.append(svgRectElem);
            $commitTableSVG.append(svgTextElem);
            const box_width = svgTextElem.getBBox().width + 10;
            // TODO: This may be causing slowdown. Use document.getElementById('#id');
            $('#' + branchRectId).attr('width', box_width);
            currentX += box_width + self.BRANCH_TEXT_SPACING;
        }

        // Draw the summary text.
        const entryElem = self.makeSVG('text', {x: currentX, y: pixelY + self.TEXT_Y_ALIGNMENT, fill: 'white'});
        entryElem.textContent = self.summary;
        entryElem.oncontextmenu = contextFunction;
        $commitTableSVG.append(entryElem);
        self.width = currentX + entryElem.getBBox().width;

        const rectElem = self.makeSVG('rect', {class: 'backgroundRect', x: pixelX, y: pixelY + self.RECT_Y_OFFSET, width: self.width, height: self.RECT_HEIGHT, style: 'fill:white;fill-opacity:0.1;'});
        rectElem.oncontextmenu = contextFunction;
        $commitTableSVG.append(rectElem);
    }

    /**
     * Makes an SVG element
     * @param {string} tag
     * @param {Object<string, number|string>} attrs
     * @return {SVGElement|SVGGraphicsElement}
     */
    makeSVG(tag, attrs) {
        const el = document.createElementNS('http://www.w3.org/2000/svg', tag);
        // eslint-disable-next-line guard-for-in
        for (const k in attrs) {
            el.setAttribute(k, attrs[k]);
        }
        return el;
    }

    /**
     * Gets the color of the row based on the indent
     * @param {number} xValue
     * @return {string}
     */
    getColor(xValue) {
        const colorNum = xValue % 4;
        if (colorNum === 0) {
            return '\#00CC19';
        } else if (colorNum === 1) {
            return '\#0198A6';
        } else if (colorNum === 2) {
            return '\#FF7800';
        } else {
            return '\#FF0D00';
        }
    }

    /**
     * Gets the function to be called by oncontextmenu
     * @return {(function(*): void)|*}
     */
    getContextFunction() {
        const self = this;
        return function(event) {
            event.preventDefault();
            const $contextMenu = $('#contextMenu');
            $contextMenu.empty();
            $contextMenu.css('left', event.pageX + 'px');
            $contextMenu.css('top', event.pageY + 'px');

            const $mergeBtn = $('<button type="button" class="btn btn-outline-light btn-sm rounded-0 cm-item"><i class="bi bi-arrows-angle-contract"></i> Merge</button>');
            $mergeBtn.click(function() {
                // TODO: Implement stuff here
            });
            $contextMenu.append($mergeBtn);

            const $cherrypickBtn = $('<button type="button" class="btn btn-outline-light btn-sm rounded-0 cm-item"><i class="bi bi-bullseye"></i> Cherrypick Commit</button>');
            $cherrypickBtn.click(function() {
                // TODO: Implement stuff here
            });
            $contextMenu.append($cherrypickBtn);

            const $copyShaBtn = $('<button type="button" class="btn btn-outline-light btn-sm rounded-0 cm-item"><i class="bi bi-clipboard"></i> Copy SHA</button>');
            $copyShaBtn.click(function() {
                writeText(self.sha).then();
            });
            $contextMenu.append($copyShaBtn);

            const $softResetBtn = $('<button type="button" class="btn btn-outline-light btn-sm rounded-0 cm-item"><i class="bi bi-arrow-clockwise"></i> Soft Reset to Here</button>');
            $softResetBtn.click(function() {
                // TODO: Implement stuff here
            });
            $contextMenu.append($softResetBtn);

            const $mixedResetBtn = $('<button type="button" class="btn btn-outline-light btn-sm rounded-0 cm-item"><i class="bi bi-arrow-clockwise"></i> Mixed Reset to Here</button>');
            $mixedResetBtn.click(function() {
                // TODO: Implement stuff here
            });
            $contextMenu.append($mixedResetBtn);

            const $hardResetBtn = $('<button type="button" class="btn btn-outline-light btn-sm rounded-0 cm-item"><i class="bi bi-arrow-clockwise"></i> Hard Reset to Here</button>');
            $hardResetBtn.click(function() {
                // TODO: Implement stuff here
            });
            $contextMenu.append($hardResetBtn);

            $contextMenu.show();
        };
    }
}
